OrbitLinuxTracing/TracerThread.cpp

```diff
@@ -92,28 +103,48 @@ void TracerThread::InitUprobesEventProcessor() {
 bool TracerThread::OpenUprobes(const std::vector<int32_t>& cpus) {
   bool uprobes_event_open_errors = false;
   absl::flat_hash_map<int32_t, int> uprobes_ring_buffer_fds_per_cpu;
+  absl::flat_hash_map<int32_t, std::vector<int>> fds_per_cpu;
 
   for (const auto& function : instrumented_functions_) {
     absl::flat_hash_map<int32_t, int> function_uprobes_fds_per_cpu;
     absl::flat_hash_map<int32_t, int> function_uretprobes_fds_per_cpu;
     bool function_uprobes_open_error = false;
 
+    uint64_t virtual_address = function.VirtualAddress();
+    bool open_uprobe = true;
+    bool open_uretprobe = true;
+
+    // Manual instrumentation relies on two distinct function calls, start and
+    // stop. Only open uprobe on start and uretprobe on stop.
+    if (manual_instrumentation_config_.IsTimerStartAddress(virtual_address)) {
+      open_uretprobe = false;
+    } else if (manual_instrumentation_config_.IsTimerStopAddress(
+                   virtual_address)) {
+      open_uprobe = false;
+    }
+
     for (int32_t cpu : cpus) {
-      int uprobes_fd = uprobes_retaddr_event_open(
-          function.BinaryPath().c_str(), function.FileOffset(), -1, cpu);
-      if (uprobes_fd < 0) {
-        function_uprobes_open_error = true;
-        break;
+      if (open_uprobe) {
+        int uprobes_fd = uprobes_retaddr_event_open(
+            function.BinaryPath().c_str(), function.FileOffset(), -1, cpu);
+        if (uprobes_fd < 0) {
+          function_uprobes_open_error = true;
+          break;
+        }
+        function_uprobes_fds_per_cpu.emplace(cpu, uprobes_fd);
+        fds_per_cpu[cpu].emplace_back(uprobes_fd);
```

Comment:
If (for some reason I'm not aware of, but bear with me) this uprobes fails to open not on the first cpu but on a subsequent one, all file descriptors for this function get closed (in the `if (function_uprobes_open_error)`), but don't you end up having recorded the file descriptor in `fds_per_cpu` (which you use later for the redirection) and that stays there? Or am I understanding incorrectly?

Reply:
You're right. I completely rewrote the loop that creates uprobes and uretprobes. I think it's easier to reason about now. Please let me know what you think.